### PR TITLE
fix: Authorisation before Creation & Deletion API Key

### PR DESCRIPTION
**Pull request by ShubhamPalriwala**
## What does this PR do?
Validates the user is authorized enough to delete the API Key:
- Gets the user details from nextjs server session
- Based on the ApiKey provided from CLIENT-SIDE, we query the DB again to get the Api Key
    - We query it again to get the environment ID of the API Key from the server itself rather than the client side (query params which can be modified)
- We parse the environment ID from the queried API Key and check if the user has access to the environment
- Only now is the user allowed to delete the API Key

PS: I also found a duplicated DB method which was unused so modified it to get ApiKey from ID!

The only thing client side here is the ApI Key ID!

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist
- [ ] Added a screen recording or screenshots to this PR
- [ ] Filled out the "How to test" section in this PR
- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: packages/lib/services/apiKey.ts
Comment (mattinannt) [inline packages/lib/services/apiKey.ts:11]: why did you rename this? I strongly prefer “getApiKey” as it’s the default method to get it by id

Comment (mattinannt) [reply packages/lib/services/apiKey.ts:11]: also is this rename/remove of thd old function affecting existing functionality?

File: apps/web/lib/authorization.ts
Comment (mattinannt) [inline apps/web/lib/authorization.ts:1]: Please move to `@formbricks/lib` as we would like to have all library functions there in the future 😊

File: apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts:11]: is it maybe possible to have a generic method `canUserAccessRessource` that works not only for ApiKeys but also for e.g. surveys? Any ideas? 😊
If not possible please move the authorization method to the api key service.

**Review by mattinannt — COMMENTED** [review]
@ShubhamPalriwala thanks a lot for the first draft 💪. I made a few comments of my first thoughts. After you addressed them, please share the updated version in the #dev channel in slack, briefly (very short) explain your thoughts behind this and ask for feedback from the other developers 😊💪

File: packages/lib/services/apiKey.ts
Comment (ShubhamPalriwala) [reply packages/lib/services/apiKey.ts]: renamed! also, no it does not affect existing functionality because we do not use this method anywhere else! and this method was similar to the `getApiKeyFromKey()` which still exists!

File: apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts
Comment (ShubhamPalriwala) [reply apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts:11]: tried the generic method but it wasn't looking promising and each resource would have a very different take! hence deciding to keep them all separate!

File: apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts:11]: Please use `isAuthorized`. I think we should make it consistent from now on prefixing boolean variables with `is` :-)

File: apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts:10]: should we also add a `if (!session) {throw ...}` here?

And do you need the explicit type of session? Isn't `getServerSession` giving this already?

File: packages/lib/services/apiKey.ts
Comment (mattinannt) [inline packages/lib/services/apiKey.ts:10]: that import seems wrong. we should move the apiHelper to the lib package.

Comment (mattinannt) [reply packages/lib/services/apiKey.ts:10]: Please move to `packages/lib/helpers/api`. Currently still thinking about the notation. I will ask in Slack.

File: packages/lib/services/apiKey.ts
Comment (mattinannt) [inline packages/lib/services/apiKey.ts:117]: Please add type for Session

File: apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts:23]: If we now use `session` for the new methods instead of `session.user` we should also change it in `hasUserEnvironmentAccess` to avoid confusion

Comment (ShubhamPalriwala) [reply apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts]: Thought about this and decided to use **user** across the methods because we are validating the user itself hence that only should be passed rather than the session object!

**Comment by mattinannt** [comment]
Putting this on hold waiting for the new authentication logic to get merged first.

File: packages/lib/environment/auth.ts
Comment (mattinannt) [inline packages/lib/environment/auth.ts:4]: please use `userId` instead of the user object to make it better reusable for different occasions.

Comment (mattinannt) [inline packages/lib/environment/auth.ts:4]: can we add a cache to this? maybe no revalidationTags, just 5 minute revalidation period?

Comment (ShubhamPalriwala) [reply packages/lib/environment/auth.ts]: done

Comment (ShubhamPalriwala) [reply packages/lib/environment/auth.ts]: done

File: packages/lib/environment/auth.ts
Comment (mattinannt) [inline packages/lib/environment/auth.ts:33]: I think we can cache the environment directly. Please add unstable_cache directly to the hasUserEnvironmentAccess.

File: packages/lib/environment/auth.ts
Comment (mattinannt) [inline packages/lib/environment/auth.ts:38]: please change key to `users-${userId}-environments-${environmentId}`
and also add the environment as a tag `environments-${environmentId}`

File: apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/settings/api-keys/actions.ts:12]: Please throw a `AuthorizationError` from `@formbricks/types/v1/errors`
Same in other places where you throw an error.

**Review by mattinannt — CHANGES_REQUESTED** [review]
@ShubhamPalriwala Looks great, just a few smaller things before we can merge :-)

File: packages/lib/environment/auth.ts
Comment (mattinannt) [inline packages/lib/environment/auth.ts:5]: if you return the function directly you can instead of
```
export const myFunc = (userId: string) => { return null }
```
can also write 
```
export const myFunc = (userId: string) => null
```

But we can merge, just wanted to mention this ;-)

**Review by mattinannt — APPROVED** [review]

File: packages/lib/environment/auth.ts
Comment (ShubhamPalriwala) [reply packages/lib/environment/auth.ts]: done

File: packages/lib/environment/auth.ts
Comment (ShubhamPalriwala) [reply packages/lib/environment/auth.ts:5]: yeah my bad, will update this next time!

**Pull request merged**
